[models] Add AudioFlamingo3 integration

lashahub opened the pull request:
This PR adds support for **AudioFlamingo3 (AF3)** — NVIDIA’s open large audio language model capable of reasoning over speech, sounds, and music.

- **Paper**: [Audio Flamingo 3: Advancing Audio Intelligence with Fully Open Large Audio Language Models](https://arxiv.org/abs/2507.08128)
- **Model Weights**: [nvidia/audio-flamingo-3 on Hugging Face](https://huggingface.co/nvidia/audio-flamingo-3)
- **Original Code**: [NVIDIA Audio Flamingo GitHub (branch: audio_flamingo_3)](https://github.com/NVIDIA/audio-flamingo/tree/audio_flamingo_3)

It introduces the following components:

- `AudioFlamingo3` model class
- `AudioFlamingo3Processor` for preprocessing text + audio
- Configuration, modeling, and processing utilities
- Example usage

With this integration, AF3 can be loaded directly from the Hugging Face Hub:

```python
from transformers import AudioFlamingo3Processor, AudioFlamingo3

processor = AudioFlamingo3Processor.from_pretrained("nvidia/audio-flamingo-3")
model = AudioFlamingo3.from_pretrained("nvidia/audio-flamingo-3")

prompt = "What is happening in the audio?"
audio = "clap.wav"

input_ids, media, media_meta = processor(prompt, audio)
output_ids = model.generate(
    input_ids=input_ids,
    media=media,
    media_meta=media_meta,
    generation_config=model.default_generation_config,
)
print(processor.decode(output_ids))
# Example output: "A crowd is applauding and cheering."
```

lashahub committed:
Audio Flamingo 3 initial integration

lashahub committed:
Added local Qwen

lashahub committed:
Moving to AF3

lashahub committed:
Loading directly from HF

lashahub committed:
Formatting

lashahub committed:
Merge branch 'huggingface:main' into main

lashahub committed:
Merge branch 'main' of https://github.com/jsalt-2025/transformers

lashahub committed:
add snapshot_download

lashahub committed:
Loading from hub

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

lashahub committed:
Import gating

lashahub committed:
Merge branch 'audioflamingo3' of https://github.com/jsalt-2025/transformers into audioflamingo3

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

ebezzam reviewed (changes requested):
Thanks @lashahub for your PR! This is very a exciting model to add to the Transformers library!

I see that you've taken inspiration from [Llava](https://github.com/huggingface/transformers/tree/main/src/transformers/models/llava), which makes sense as you combine modules of different modalities.

Most of my comments are about rearranging your modules so that they fit the Transformers convention, such that it will be more convenient for others to use and to test your new model. To that end, [Llava](https://github.com/huggingface/transformers/pull/27662/files)'s and [this PR](https://github.com/huggingface/transformers/pull/38405/files) (of another audio model) might serve as a useful example to see what files will be added/modified.

Below are my suggested steps.

## 1. Refactoring / reorganizing current files according to Transformers convention

You can take inspiration from the above models for refactoring your `configuration`, `modeling` and `processing` files, specifically:
- Consolidating your configurations. From what I see you may need only one config like in [Llava](https://github.com/huggingface/transformers/blob/f0e87b436d8bd4dd4225da235ac8f988d4bbe845/src/transformers/models/llava/configuration_llava.py#L24) or two like in [Dia](https://github.com/huggingface/transformers/blob/main/src/transformers/models/dia/configuration_dia.py) for `AudioFlamingo3Config` and `AudioFlamingo3EncoderConfig`.
- Processor: you can take inspiration from [Dia](https://github.com/huggingface/transformers/blob/f0e87b436d8bd4dd4225da235ac8f988d4bbe845/src/transformers/models/dia/processing_dia.py#L62) to group your feature extractor, text tokenizer, and audio tokenizer into a single component. Loading pre-trained feature extractors and LLMs can be directly handled by the processor **without** you (or the user) having to manually load the corresponding weights (like you've done below).
```python
# Load components
fe = WhisperFeatureExtractor.from_pretrained("openai/whisper-large-v3")
tok = AutoTokenizer.from_pretrained(
    llm_dir,
    padding_side="right",
    use_fast=True,
    legacy=False,
)
```
To this end, you'll need to create a model conversion script (with something like [this](https://github.com/huggingface/transformers/blob/29ddcacea3ad9d3cdf6c5d8e51d1d39cbc5e7dfa/src/transformers/models/dia/convert_dia_to_hf.py#L191)) so the configuration files are generated for the processor to know where to pull the relevant models.
- The `modeling` file will diminish quite significantly because we won't apply the audio/text tokenizer inside it but rather in the processor, and the configuration file will also handle pulling the relevant LLM config.

## 2. Conversion script

This will be needed to convert your model weights / configuration to one that is compatible with the one defined in the above files.

This script can also handle uploading the Transformer compatible model and its configuration to the Hugging Face Hub. You can again take inspiration from [Llava](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/convert_llava_weights_to_hf.py) and [Dia](https://github.com/huggingface/transformers/blob/main/src/transformers/models/dia/convert_dia_to_hf.py).

## 3. Testing, documentation, etc

Once your model implementation is consistent with other models implemented in Transformers, there's a lot of boilerplate code we can reuse to make using model convenient and to apply various testing suites. For example, you can see the `docs`, `src/transformers/models/auto`, and `tests/models/dia` folders of the [Dia PR](https://github.com/huggingface/transformers/pull/38405/files#diff-033858e553e3b75b892806499aa57be7cd1c66c61b292f8ce0a3060a4b31ea64) on how to prepare / modify the relevant files.

Hope that helps and let me know if you have any questions!

lashahub committed:
Pass audio arrays directly

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

lashahub committed:
Merge branch 'audioflamingo3' of https://github.com/jsalt-2025/transformers into audioflamingo3

lashahub committed:
Remove requires_backend

lashahub committed:
Move constants to config.json

lashahub committed:
Remove redundancies

lashahub committed:
Separate tokenizer, cleaner from_pretrained

lashahub committed:
Remove LlavaMetaModel

lashahub committed:
Remove sound tower wrapper

lashahub committed:
Merged BasicSoundEncoder

lashahub committed:
Some improvements

lashahub committed:
Towards AudioFlamingo3

lashahub committed:
Migrate LlavaConfig

lashahub committed:
Merge LlavaMetaForCausalLM into AudioFlamingo3ForConditionalGeneration

lashahub committed:
Remove redundant lines

lashahub committed:
Add AudioFlamingo3PreTrainedModel

lashahub committed:
Unified model.safetensors

lashahub committed:
Inline MM projector

lashahub committed:
Tokenizer in root dir

lashahub committed:
Default processor from_pretrained

lashahub committed:
Remove tokenizer from modeling

lashahub committed:
Added types

lashahub committed:
Cleanup

lashahub committed:
Docs & license

lashahub committed:
device handling

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

lashahub committed:
Merge branch 'audioflamingo3' of https://github.com/jsalt-2025/transformers into audioflamingo3

ebezzam reviewed (commented):
@lashahub Thanks for the changes! It's already looking much cleaner 👏 

I've given more comments to improve compatibility, structure, and style according to Transformers.

Regarding your open question, keeping a separate `AudioFlamingo3EncoderConfig` as you've done looks great 👍 

As for next steps, what you've mentioned (conversion script, tests, and docs) is perfect! I would also add (and addressing below first):
- Writing a [modular script](https://huggingface.co/docs/transformers/en/modular_transformers) to make best use of other components in Transformers. I've tried to give pointers on how to best to do that.
- I think a `feature_extraction_audioflamingo3.py` will be necessary to define `AudioFlamingo3FeatureExtractor`, as it seems that `WhisperFeatureExtractor` doesn't prepare audio inputs as is needed for your model?
- Cleaning up your `_embed` function to be more consistent with [Qwen2](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L517) and ensuring that attention masks are properly prepared in the processor.

Great job so far, and let me know if you have any questions!

lashahub committed:
Change year

lashahub committed:
Remove redundant methods

lashahub committed:
Use BatchFeature

lashahub committed:
Streamline audio feature handling

lashahub committed:
Batch inference

lashahub committed:
Reorder alphabetically

lashahub committed:
Make style check

lashahub committed:
Make fixup

lashahub committed:
Avoid calls to separate functions

lashahub committed:
Remove forward_tower()

lashahub committed:
Rename encode_sound to get_audio_features for clarity

lashahub committed:
Add batch decoding method to AudioFlamingo3Processor

lashahub committed:
Use tensors instead of lists

lashahub committed:
Move end embed token eval

lashahub committed:
Prepare audio_features_mask in the processor

lashahub committed:
No hardcoded 750 and 3000

lashahub committed:
Remove _load_sound_mask completely and use WhisperFeatureExtractor

lashahub committed:
Compute embeddings separately

lashahub committed:
MM Projector is audio adaptor

lashahub committed:
Simplify AudioFlamingo3Config initialization with default encoder_config

lashahub committed:
Add modular

lashahub committed:
Clean up

lashahub committed:
make fixup

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,276 @@
+        fe_kwargs = dict(output_kwargs.get("audio_kwargs", {}))
+        fe_kwargs.update(
+            {
+                "return_attention_mask": True,
+                "padding": "max_length",
+                "truncation": True,
```
can you put these args inside `audio_kwargs` like in [Dia](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/dia/processing_dia.py#L50C10-L50C22)

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,276 @@
+                "return_attention_mask": True,
+                "padding": "max_length",
+                "truncation": True,
+                "return_tensors": "pt",
```
this can be in `common_kwargs` like in [Dia](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/dia/processing_dia.py#L58)

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,276 @@
+            prompts,
+            padding=output_kwargs["text_kwargs"].get("padding", True),
+            truncation=output_kwargs["text_kwargs"].get("truncation", True),
+            return_tensors="pt",
```
as for Feature extraction, `return_tensors` can be obtained from `common_kwargs`

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,276 @@
+
+        return self.tokenizer(text, return_tensors="pt").input_ids[0]
+
+    def _get_num_windows(self, T: int, sr: int) -> tuple[int, int]:
```
if not used anymore, we can delete. Or did you want to use within `__call__`?

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,276 @@
+            if T <= wl:
+                num_windows = 1
+            elif T >= (20 * wl - 19 * wo):  # Limit to maximum of 20 windows
+                num_windows = 20
```
is there a reason it can't go more than 20? 

if this is the limit, how about a warning in case someone wants to work with audio longer than this?

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,276 @@
+        )
+
+        fe = self.feature_extractor
+        sr = int(getattr(fe, "sampling_rate", 16_000))
+        hop = int(getattr(fe, "hop_length", 160))
+        n_samples = int(getattr(fe, "n_samples", int(30.0 * sr)))
```
Can these be added to `audio_kwargs`? Namely 16000, 160, and 30.

It will help to see the parameters clearly presented above.

See [Dia](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/dia/processing_dia.py#L56) which also adds the sampling rate

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,512 @@
+        return inputs_embeds, labels, attention_mask
+
+    @torch.inference_mode()
+    def generate(
```
Instead of overwriting the `generate` method, we typically define the [`forward`](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L715) method, and overwrite [`prepare_inputs_for_generation`](https://github.com/huggingface/transformers/blob/37103d6f22cd25c9eba7297ae79ed8c47a852e54/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L882). I put example links to Qwen.

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,512 @@
+        return self.llm.generate(inputs_embeds=inputs_embeds, attention_mask=attn, **generation_kwargs)
+
+    @property
+    def default_generation_config(self) -> GenerationConfig:
```
This is not something we normally have.

I saw that in your most [recent example](https://github.com/huggingface/transformers/pull/40290#issuecomment-3269211209), you're using this to pass the token IDs.

This is something that can be handled if you overwrite  `prepare_inputs_for_generation` as mentioned above.

Moreover, like [Dia](https://github.com/huggingface/transformers/blob/8a091cc07c9808196e3154ab8bd6252d919f929b/src/transformers/models/dia/configuration_dia.py#L340-L342) and [Llava](https://github.com/huggingface/transformers/blob/91be12bdc6d312a2d7a44174971f025c4023a91f/src/transformers/models/llava/configuration_llava.py#L87), let's have the token IDs in the configuration.

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,276 @@
+        encoder_mask_len = tokens_per_window_max + 1  # Reserve one additional slot for end token
+
+        # Audio windowing configuration: 30 second windows with no overlap by default
+        wl = n_samples
+        wo = 0
```
Can we just use `n_samples`? and use a more descriptive name than `wo`?

ebezzam reviewed (commented):
@lashahub this is coming together very nicely! Thanks for switching to modular and updating the processor. I've left some comments to better conform to Transformers-convention.

We may still come back to `_merge_input_ids_with_audio_features` if a core maintainer thinks we should move some of the processing to `AudioFlamingo3Processor`, but we can leave it like this for now.

For next steps, how about:
- Adding the conversion script which pushes a model to the hub.
- Starting the unit tests, first with integration tests to make sure that the Transformers-compatible model is consistent with the original model. For example, [here](https://gist.github.com/ebezzam/6382bdabfc64bb2541ca9f77deb7678d) you can see scripts I created to test another Nvidia model to write the expected outputs to a JSON file. And [here](https://github.com/nithinraok/transformers/blob/d04760a7f8a54f60820ec577c0eb229413e8b027/tests/models/parakeet/test_modeling_parakeet.py#L310) are the tests themselves.
- Start the markdown doc file

Let me know if you have any questions or need any pointers!

lashahub committed:
Cleanup processing, add params to encoder config

lashahub committed:
Remove redundant methods

lashahub committed:
update config references, improve method names, and enhance logging in processor

lashahub committed:
processor: move FE args to audio_kwargs, use common_kwargs for return_tensors

lashahub committed:
Qwen-like processor

lashahub committed:
Simplified AudioFlamingo3Processor

lashahub committed:
Extract common code from generate() and forward()

lashahub committed:
Add conversion script for AudioFlamingo3 to Hugging Face format

lashahub committed:
Use save_pretrained()

lashahub committed:
Don't overwrite gen config

lashahub committed:
Use AutoTokenizer and FE to convert the processor

lashahub committed:
minor formatting

lashahub committed:
Finalize processor, do token expansion inside

lashahub committed:
AudioFlamingo3: refactor docs, types, and audio–text feature merge

lashahub committed:
AudioFlamingo3 Docs

lashahub committed:
Add AudioFlamingo3Processor to AutoProcessor

lashahub committed:
Processor tests

lashahub committed:
Use audio_config instead of encoder_config

lashahub committed:
Add audio_token_id to config

lashahub committed:
Cleanup & new keys

lashahub committed:
Add links

lashahub committed:
Improved processor

lashahub committed:
Handle conversational input

ebezzam committed:
Make processing consistent.

ebezzam committed:
Merge branch 'audioflamingo3' of github.com:jsalt-2025/transformers into audioflamingo3

ebezzam committed:
Add fallback for no sound token, default left padding.

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,214 @@
+                    if user_start in sample:
+                        # Chat template: insert after user start
+                        sample = sample.replace(user_start, user_start + prefix, 1)
+                    else:
+                        # Plain text: prepend to the beginning
+                        sample = prefix + sample
```
@lashahub Added this fallback as you have in your docs:

`"You can omit `<sound>` entirely and let the processor prepend the expanded tokens:" `

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,214 @@
+    _defaults = {
+        "text_kwargs": {
+            "padding": True,
+            "padding_side": "left",
```
@lashahub In all your examples, you're setting `padding_side=left`. How about defaulting to this?

ebezzam reviewed (commented):
@lashahub Thanks for your changes, I made a few tweaks to the processor. 

Before I saw you were using the chat template within the processor and then removed it. We indeed don't want to use chat template within the processor, but _we do_ want to be able to support such usage! Below is an example of how we could do that (in a batch):
```python
# Create two separate conversations instead of one conversation with two user messages
conversations = [
    [  # First conversation
        {"role": "system", "content": [{"type": "text", "text": "You are a helpful assistant."}]},
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "Transcribe the input speech."},
                {
                    "audio": load_audio("src/transformers/models/audioflamingo3/speech.wav"),
                    "type": "audio",
                },
            ],
        },
    ],
    [  # Second conversation
        {"role": "system", "content": [{"type": "text", "text": "You are a helpful assistant."}]},
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "Describe the music in details."},
                {
                    "audio": load_audio("src/transformers/models/audioflamingo3/music_16khz.wav"),
                    "type": "audio",
                },
            ],
        },
    ],
]
chat_template = """{% set audio_count = namespace(value=0) %}{% for message in messages %}<|im_start|>{{ message['role'] }}
{% if message['content'] is string %}{{ message['content'] }}{% else %}{% set audio_tokens = namespace(value='') %}{% set text_tokens = namespace(value='') %}{% for content in message['content'] %}{% if content['type'] == 'audio' or 'audio' in content %}{% set audio_count.value = audio_count.value + 1 %}{% if 'path' in content %}{% set audio_tokens.value = audio_tokens.value + '<sound>' %}{% elif 'url' in content %}{% set audio_tokens.value = audio_tokens.value + '<sound>' %}{% elif 'audio' in content %}{% set audio_tokens.value = audio_tokens.value + '<sound>' %}{% elif 'array' in content %}{% set audio_tokens.value = audio_tokens.value + '<sound>' %}{% endif %}{% elif 'text' in content %}{% set text_tokens.value = text_tokens.value + content['text'] %}{% endif %}{% endfor %}{{ audio_tokens.value }}{{ text_tokens.value }}{% endif %}<|im_end|>
{% endfor %}{% if add_generation_prompt %}<|im_start|>assistant
{% endif %}"""

# Test with processor but without tokenization first
inputs = processor.apply_chat_template(
    conversations,
    chat_template=chat_template,
    tokenize=True,
    add_generation_prompt=True,
    padding_side="left",
    return_dict=True,
).to(device)

output_ids = model.generate(**inputs, max_new_tokens=2048, do_sample=False)
texts = processor.batch_decode(output_ids, skip_special_tokens=True)
texts = [text.split("\nassistant\n")[-1] for text in texts]
print(texts)
```
When you removed the chat template within the processor, I think you removed support for when there is no `<sound>` token, so I've added that.

I've also given some tips to update the conversion scripts, to configure the Jinja chat template (so you don't need to manually set up like above) and to automate uploading to the Hub.

Let me know when you make the changes and add the modeling tests, and then we can soon get a core maintainer to take a look!

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
````diff
@@ -0,0 +1,159 @@
+print(texts)
+```
+
+Tips:
+
+* Left padding is recommended for batched generation with mixed prompt lengths.
+* The processor returns conversation-formatted text; the snippet above keeps only the assistant’s reply.
````
Remove? With left padding as default + chat_template removed from within the processor

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
```diff
@@ -0,0 +1,159 @@
+### Architecture
+
+* **AudioFlamingo3Encoder**
+  Whisper-style conv front-end → Transformer encoder → average-pool over time (stride 2) → LayerNorm.
```
"Whisper-style conv front-end → Transformer encoder" --> "Whisper-style feature extractor + encoder"

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
Examples can remove `padding_side="left"` if we keep the default in the processor.

`tensor_type` can also be removed (as I've also set to default to `pt`), or renaming to `return_tensors` which is the name typically used in Transformers (see [Qwen example](https://huggingface.co/docs/transformers/v4.56.2/en/model_doc/qwen2_audio#inference)).

lashahub committed:
Cleanup

lashahub committed:
Replace manual 4D mask with masking_utils; dtype/device from inputs

lashahub committed:
Text only mode

lashahub committed:
Finalize processor

lashahub committed:
Export processor directly

lashahub committed:
Add push_to_hub to converter

lashahub committed:
Add model_input_names property to AudioFlamingo3Processor to pass tests

lashahub committed:
Processor chat template support

lashahub committed:
Added Jinja processor chat template with audio support

lashahub committed:
Processor tests

lashahub committed:
Model tests

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

lashahub committed:
Added docs

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

lashahub committed:
Don't use common_kwargs in __call__

lashahub committed:
Pass 'test_left_padding_compatibility' by never treating padding as content

lashahub committed:
Updated docs

lashahub committed:
Cleanup docs

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
````diff
@@ -0,0 +1,165 @@
+
+## Quickstart
+
+```python
````
How about just a short text above explaining the example snippet? (Batch processing that does both transcription and music understanding)

ebezzam replied:
and another short example just above that shows just transcription with a single example. In general people like seeing multiple examples for different tasks so they can conveniently copy-paste

ebezzam commented on src/transformers/models/audioflamingo3/configuration_audioflamingo3.py:
```diff
@@ -0,0 +1,206 @@
+        encoder_ffn_dim: int = 5120,
+        encoder_layerdrop: float = 0.0,
+        activation_function: str = "gelu",
+        d_model: int = 1280,
```
let's rename to `hidden_size` like other/recent Transformers models. You can see a recent model addition ([Voxtral](https://github.com/huggingface/transformers/blob/main/src/transformers/models/voxtral/configuration_voxtral.py#L35)) for typical naming for a similar model

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,517 @@
+
+    def __init__(self, config: AudioFlamingo3Config) -> None:
+        super().__init__()
+        d_audio = config.audio_config.d_model
```
reminder: to update to `hidden_size` and other changes to config names to make consistent with other Transformer models

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,517 @@
+    def __init__(self, config: AudioFlamingo3EncoderConfig):
+        super().__init__(config)
+        self.avg_pooler = nn.AvgPool1d(config.avg_pool_kernel_size, stride=config.avg_pool_stride)
+        self.post_init()
```
Just noticed that `Qwen2AudioEncoder`  is like yours, albeit with hardcoded values to `AvgPool1d`:

https://github.com/huggingface/transformers/blob/39b6d3bf7e30b92b2de50a31ee991557d41ab568/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L322

Anyway let's use that with modular:
```python
class AudioFlamingo3Encoder(Qwen2AudioEncoder)
...
```
So that you don't need to overwrite `__init__`.

You can then remove the following from your config:
```python
self.avg_pool_kernel_size = avg_pool_kernel_size
self.avg_pool_stride = avg_pool_stride
```
(Or add them to that of [Qwen2AudioEncoderConfig](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_audio/configuration_qwen2_audio.py#L24C7-L24C30) if you think it's important to expose them.)

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,517 @@
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        input_features: Optional[torch.FloatTensor] = None,  # (#windows, n_mels, T_mel)
```
_note to self: change to `audio_spectrogram` for more intuitive input as per standardization for audio models?_

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,517 @@
+            post = enc_out.last_hidden_state  # (#windows, S_out, C)
+            audio_feats = self.multi_modal_projector(post)  # (#windows, S_out, D)
+
+            N, S_out_max, D = audio_feats.shape
```
`S_out_max = audio_feats.shape[1]` is enough?
 
Convention on our side to avoid single letter variable and unused ones

ebezzam commented on tests/models/audioflamingo3/test_modeling_audioflamingo3.py:
```diff
@@ -0,0 +1,421 @@
+                        raise ValueError("The eager model should not have SDPA attention layers")
+
+
+@require_torch
```
Let's put such decorators on methods rather than classes

ebezzam commented on tests/models/audioflamingo3/test_modeling_audioflamingo3.py:
```diff
@@ -0,0 +1,421 @@
+        urls = [
+            "https://audioflamingo3.github.io/static/emergent/Dogs%20barking%20in%20sync%20with%20the%20music.wav",
+            "https://audioflamingo3.github.io/static/emergent/Aside%20from%20_Interstellar%20Spaces_,%20Saxophones%20do%20not%20bark%20nor%20meow.wav",
+            "https://audioflamingo3.github.io/static/emergent/Unlikely%20existence%20of%20banjo%20and%20rain%20sounds%20at%20the%20same%20time%20in%20the%20training%20data.wav",
```
You can use these URLs in the examples I mentioned above

ebezzam commented on tests/models/audioflamingo3/test_modeling_audioflamingo3.py:
```diff
@@ -0,0 +1,421 @@
+
+
+@require_torch
+class AudioFlamingo3ForConditionalGenerationIntegrationTest(unittest.TestCase):
```
For the integration tests, what we will want to test are expected outputs for single file and batch inference against the original checkpoint.

Here are are examples for NVIDIA's Parakeet model ([single](https://github.com/huggingface/transformers/blob/7e204ad121e729cac59ea213356c68527cac0a7e/tests/models/parakeet/test_modeling_parakeet.py#L329) and [batched](https://github.com/huggingface/transformers/blob/7e204ad121e729cac59ea213356c68527cac0a7e/tests/models/parakeet/test_modeling_parakeet.py#L354)).

And also writing a reproducer like [this](https://github.com/huggingface/transformers/blob/7e204ad121e729cac59ea213356c68527cac0a7e/tests/models/parakeet/test_modeling_parakeet.py#L331C60-L331C150) to be able to reproduce the expected outputs (in case of regression) and that can write them in a JSON file that can be stored in the [`fixtures`](https://github.com/huggingface/transformers/tree/main/tests/fixtures) folder.

ebezzam commented on tests/models/audioflamingo3/test_modeling_audioflamingo3.py:
```diff
@@ -0,0 +1,421 @@
+        self.assertEqual(num_sound_tokens, total_post_frames)
+
+    @slow
+    def test_integration_batch_three_samples_processor_apply_chat_template(self):
```
see comment above, namely computing expected outputs against original checkpoint

ebezzam reviewed (commented):
@lashahub thanks for the changes! I've left some comments on:
1. making even better use of modular for the Encoder model
2. standardize with other (recent) Transformer models (namely Voxtral which is similar in architecture)
3. integration tests that compare against expected outputs of the checkpoint current on [your HF page](https://huggingface.co/nvidia/audio-flamingo-3) (and backing it up elsewhere in case you decide to overwrite the HF page with the Transformers-compatible checkpoint)

ebezzam committed:
Standardization

ebezzam committed:
Update conversion script weight mapping.

ebezzam committed:
Flatten _build_square_attn_mask

ebezzam committed:
Make style

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,457 @@
+        attention_mask: Optional[torch.Tensor] = None,
+        feature_attention_mask: Optional[torch.Tensor] = None,
+        # TODO (ebezzam) why unused? normally passed to self.language_model
+        position_ids: Optional[torch.LongTensor] = None,
```
Is there a reason to not pass to self.language model? Like in Qwen below

https://github.com/huggingface/transformers/blob/a8cece13e29057fc9712543402d73e78796b306d/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L821

ebezzam commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,457 @@
+        # Add positions, dropout
+        inputs_embeds = inputs_embeds.permute(0, 2, 1)  # (B, S_in, C)
+        # TODO (ebezzam) can `self.embed_positions.weight` be used?`
+        positions = torch.arange(inputs_embeds.shape[1], device=inputs_embeds.device).unsqueeze(0)
```
If not, we could probably also use modular to copy of the `forward` method from `Qwen2AudioEncoder`!

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
_Note to self to come back to this file after changes (e.g. variable naming, comparing to Qwen/Voxtral)._

ebezzam reviewed (commented):
@lashahub I pushed some standardization changes, and left some comments on things to still address, as well as those from the previous [review](https://github.com/huggingface/transformers/pull/40290#pullrequestreview-3350535943) on:
- updating integration tests
- updating example usage
- backing up original weights

hope that helps and thanks for your work!

lashahub committed:
Small dim and attn mask fix

lashahub committed:
Merge branch 'huggingface:main' into audioflamingo3

lashahub committed:
Fix processor padding side bug

lashahub committed:
Error handling in converter

lashahub committed:
Use position_ids

lashahub committed:
Cleanup generation config

lashahub committed:
Use precomputed position embeddings in AudioFlamingo3 encoder

lashahub committed:
Added usage examples

lashahub committed:
Fix generation config

lashahub committed:
Integration tests

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
```diff
@@ -0,0 +1,339 @@
+- Unified audio encoder across speech, sound, and music.
+- Long-audio support via windowing and post-pool alignment (up to 10 minutes).
+- Deterministic fusion that preserves sequence length by replacing `<sound>` tokens with audio embeddings.
+
```
You can add model contributions here like [this](https://github.com/huggingface/transformers/blob/2f9e3ae7f532f12ff16648fd956687354847e151/docs/source/en/model_doc/voxtral.md?plain=1#L349)

ebezzam committed:
Simplify modeling and shift part of mask preparation to processor. And update tests.

ebezzam reviewed (commented):
@lashahub thanks for the changes! 

I pushed some changes to prune code and standardize with more recent Transformers conventions (i.e. like Voxtral):
- using modular as much as possible to copy from previous models
- added a `get_audio_features` method to `AudioFlamingo3ForConditionalGeneration` and shifted some padding mask preparation to the processor
- removed output attentions logic (as it always seems to be False?)
- more descriptive variable names, e.g. `post_pool_len` and `conv_output_len` rather than `K` and `L1`
- fixed some of the tests

Please take a look and let me know if there's any breaking changes for your modeling logic. 

Could you also:
1. Address the open comments (here and in the previous review).
2. Convert and push checkpoints to HF again, to rename `init_std`: https://huggingface.co/nvidia/audio-flamingo-3-hf/blob/main/config.json#L11
3. Double check the integration tests. They are failing on our CI machine (you can see [here](https://github.com/huggingface/transformers/actions/runs/18913136148)) due to `CUDA out of memory`. Maybe you can try with smaller/different inputs? (e.g. the inputs from the [Voxtral tests](https://github.com/huggingface/transformers/blob/a43b36cf802f00616800e0bd4d748679236123ee/tests/models/voxtral/test_modeling_voxtral.py#L230))

On my machine I get the following errors:
```
# RUN_SLOW=1 pytest tests/models/audioflamingo3/test_modeling_audioflamingo3.py::AudioFlamingo3ForConditionalGenerationIntegrationTest

============================================================================================== short test summary info ===============================================================================================
FAILED tests/models/audioflamingo3/test_modeling_audioflamingo3.py::AudioFlamingo3ForConditionalGenerationIntegrationTest::test_fixture_batched_matches - RuntimeError: The frame has 0 channels, expected 1. If you are hitting this, it may be because you are using a buggy FFmpeg version. FFmpeg4 is known to fail here in some valid scenarios. Try to upgrade FFmpeg?
FAILED tests/models/audioflamingo3/test_modeling_audioflamingo3.py::AudioFlamingo3ForConditionalGenerationIntegrationTest::test_fixture_single_matches - AssertionError: The values for attribute 'shape' do not match: torch.Size([1, 21]) != torch.Size([1, 22]).
====== 2 failed, 2 warnings in 27.28s =====
```
_(although FFmpeg is probably something buggy from my environment setup)_

eustlb commented on src/transformers/models/audioflamingo3/modular_audioflamingo3.py:
```diff
@@ -0,0 +1,306 @@
+        post_lengths = (input_features_mask.sum(-1) - 2) // 2 + 1
+        valid_mask = torch.arange(audio_embeds.shape[1], device=post_lengths.device)[None, :] < post_lengths[:, None]
+        audio_embeds = audio_embeds[valid_mask.to(audio_embeds.device)]
+        return audio_embeds
```
Indeed the no-attention mask with attending to padding zeros on the 30sec fixed input window is the big limitation of Whisper encoder, and the one Moonshine adressed an improved on (along with RoPE pos encodings).

In this case, I would do something similar to what is done in [Parakeet encoder](https://github.com/huggingface/transformers/blob/37a629628329217a8f0913fd8b0119086d67058c/src/transformers/models/parakeet/modular_parakeet.py#L408-L472), meaning we handle the mask resizing directly in the audio_tower (so the audio encoder)'s forward. I do agree thought that the `attention_mask` naming becomes misleading, but I'd rather really have the mask passed to the encoder the same length as input_features passed to the encoder than doing an external resizing. This will also ease using the encoder in another model with modular if required. I would agree if possible in renaming what is here named as attention_mask in the encoder's forward to input_features_mask directly.

lashahub committed:
Updated docs

lashahub committed:
ASR convenience method

lashahub committed:
Fixed tests

lashahub committed:
make fixup

lashahub committed:
Merge branch 'main' into audioflamingo3

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
```diff
@@ -0,0 +1,393 @@
+
+## Usage patterns
+
+### Transcription shortcut
```
great, thanks for adding this example! 

@eustlb just want to double-check that this API looks good? We went for something similar to Voxtral: https://github.com/huggingface/transformers/blob/b9f90dc388fd415a2ba2a6a31a372f451d4a4eed/src/transformers/models/voxtral/processing_voxtral.py#L278

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
````diff
@@ -0,0 +1,393 @@
+print(texts)
+```
+
+➡️ transcription shortcut
````
great, thanks for adding this example! 

@eustlb just want to double-check that this API looks good? (related to offline discussion we had with Lysandre) 

We went for something similar to Voxtral: https://github.com/huggingface/transformers/blob/b9f90dc388fd415a2ba2a6a31a372f451d4a4eed/src/transformers/models/voxtral/processing_voxtral.py#L278

ebezzam committed:
Shift encoder mask preparation to the encoder's forward.

ebezzam committed:
Change to HF profiles.

ebezzam committed:
Integration test standardization.

ebezzam commented:
run-slow: audioflamingo3

ebezzam commented on tests/models/audioflamingo3/test_modeling_audioflamingo3.py:
```diff
@@ -0,0 +1,341 @@
+                        },
+                        {
+                            "type": "audio",
+                            "path": "https://audioflamingo3.github.io/static/emergent/Dogs%20barking%20in%20sync%20with%20the%20music.wav",
```
HF dataset link

ebezzam commented on tests/models/audioflamingo3/test_modeling_audioflamingo3.py:
```diff
@@ -0,0 +1,341 @@
+                        },
+                        {
+                            "type": "audio",
+                            "path": "https://audioflamingo3.github.io/static/emergent/Unlikely%20existence%20of%20banjo%20and%20rain%20sounds%20at%20the%20same%20time%20in%20the%20training%20data.wav",
```
HF dataset link

ebezzam reviewed (commented):
@lashahub thanks for the changes! I also pushed some small changes, and left some comments for some minor things to address:
- Transcription output
- Creating a mini HF dataset for the testing samples (which can also be used to store samples for Music Flamingo)

Regarding the failing tests:
- `test_eager_matches_fa2_generate` and `test_flash_attn_2_can_dispatch_composite_models` are fine on my machine and also on the [GitHub CI](https://github.com/huggingface/transformers/actions/runs/19031008789/job/54345151607) so should be fine.
- On the GitHub CI, there are some issues with loading the model for the integration tests. I'm checking internally about that. Otherwise the tests passed on my machine 👍

ydshieh commented:
run-slow: audioflamingo3

ebezzam reviewed (commented):
@eustlb A self-review for decisions made in the processor, if we can get your feedback. Thanks!

ebezzam commented on src/transformers/models/audioflamingo3/processing_audioflamingo3.py:
```diff
@@ -0,0 +1,320 @@
+                    else:
+                        logger.warning("Consider applying the chat template to ensure proper output.")
+                        # Plain text: prepend to the beginning
+                        sample = prefix + sample
```
For reference:
- Qwen2Audio: https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_audio/processing_qwen2_audio.py#L137-L164
- CSM: https://github.com/huggingface/transformers/blob/main/src/transformers/models/csm/processing_csm.py#L280-L293

ebezzam committed:
Clean up before integration test setup.

ebezzam committed:
Remove strict float32, more similar to Qwen2Audio.

ebezzam commented:
run-slow: audioflamingo3

lashahub committed:
Use HF dataset links

lashahub committed:
Keep weights in BF16

lashahub commented:
@ebezzam I wasn’t able to reproduce the test failures locally. I reinstalled the latest requirements using `uv pip install -e ".[dev-torch]"`, and both tests are passing on my end. Could you please try running them locally on your side to confirm?

ebezzam commented:
run-slow: audioflamingo3

lashahub committed:
New audio in tests

lashahub commented:
Hi @ydshieh could you run the slow tests? Thanks

ebezzam committed:
Processor conventions.

ebezzam reviewed (commented):
thanks a bunch @lashahub for your work! As mentioned below, could you also add example usage to: https://huggingface.co/nvidia/audio-flamingo-3-hf

ebezzam commented on docs/source/en/model_doc/audioflamingo3.md:
````diff
@@ -0,0 +1,393 @@
+# For a clip that may span multiple windows:
+prompt = "<sound>\nDescribe the audio in detail."
+inputs = processor(text=prompt, audio=audio_array)
+```
````
@lashahub related to [this discussion](https://github.com/huggingface/transformers/pull/40290/files#r2486281860) which I continued offline with @eustlb , I'm going to remove support for such usage, as we prefer usage through the standardized chat template.

ebezzam committed:
Standardize audio token expansion in processor.

lashahub committed:
Add 'strip_prefix' to batch_decode

lashahub commented:
Hi @ebezzam, I’ve updated the [model page](https://huggingface.co/nvidia/audio-flamingo-3-hf) with usage examples and added a `strip_prefix` parameter to `AudioFlamingo3Processor.batch_decode` (set to `False` by default) for ASR.

ebezzam committed:
Merge branch 'main' into audioflamingo3

ebezzam committed:
Batch decode nits.

ebezzam committed:
Remove dtype casting.

ebezzam commented:
@ArthurZucker could we get your review on this audio-text-to-text model? (Whisper encoder to Qwen2 LLM)
Model page: https://huggingface.co/nvidia/audio-flamingo-3-hf

Similar models which are used in modular:
- Voxtral ([model](https://huggingface.co/mistralai/Voxtral-Mini-3B-2507), [code](https://github.com/huggingface/transformers/tree/main/src/transformers/models/voxtral))
- Qwen2Audio ([model](https://huggingface.co/Qwen/Qwen2-Audio-7B-Instruct), [code](https://github.com/huggingface/transformers/tree/main/src/transformers/models/qwen2_audio))

A difference you'll see in modular (which is a nice aspect here), this model accounts for the attention mask in the encoder. See discussion [here](https://github.com/huggingface/transformers/pull/40290#discussion_r2473767582) with @eustlb

Dong-mel commented:
Hi @lashahub , thanks to the open-source model and code support!

However, after installing transformers using pip
```
pip install --upgrade git+https://github.com/huggingface/transformers
```
I found that the current version does not support AudioFlamingo3ForConditionalGeneration.
```
>>> from transformers import AudioFlamingo3ForConditionalGeneration, AutoProcessor
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'AudioFlamingo3ForConditionalGeneration' from 'transformers' (/opt/conda/lib/python3.11/site-packages/transformers/__init__.py)
```
Does it require a specific branch?

Dong-mel commented:
> Hi @Dong-mel , it hasn’t been merged into the main branch yet. You can currently use the [audioflamingo3 branch](https://github.com/jsalt-2025/transformers/tree/audioflamingo3) from the JSALT repo instead. Hopefully, it’ll be merged within the next few days.

Got it, thank you!

lashahub committed:
Read token ids from tokenizer

eustlb commented on src/transformers/models/voxtral/modular_voxtral.py:
```diff
@@ -86,7 +86,7 @@ def forward(
         if input_features.shape[-1] != expected_seq_length:
             raise ValueError(
-                f"Qwen2Audio expects the mel input features to be of length {expected_seq_length}, but found {input_features.shape[-1]}. Make sure to pad the input mel features to {expected_seq_length}."
+                f"Voxtral expects the mel input features to be of length {expected_seq_length}, but found {input_features.shape[-1]}. Make sure to pad the input mel features to {expected_seq_length}."
```
nice catch

eustlb reviewed (commented):
Great work guys, thanks a lot for the contribution! 👏
LGTM but still few things, the only truly blocking one is the `input_feature_mask` that is not one! I am going to push directly few changes in order to shortcut the process and I'll approve right after.

eustlb reviewed (commented):
forgot this in the former review

eustlb committed:
diverse changes according to review

eustlb committed:
add training example

eustlb reviewed (approved):
LGTM! 🎉

ebezzam committed:
Add missing docstring.

ebezzam committed:
Fix typos.

ebezzam committed:
Merge branch 'main' into audioflamingo3

ebezzam committed:
Add audio token docstring.

ebezzam committed:
Fix fill type.

lashahub committed:
Fix docs

lashahub committed:
Save converted weights in bf16

lashahub committed:
Fix tests

ebezzam committed:
Keep model in bf16 for tests.

ebezzam commented:
run-slow: audioflamingo3

ebezzam committed:
Update expected results for single.

ArthurZucker reviewed (approved):
Kudos everyone and thanks @eustlb and @ebezzam very clean 😉

ebezzam committed:
Fix integration tests from runner.

ebezzam commented:
run-slow: audioflamingo3

ebezzam committed:
Merge branch 'main' into audioflamingo3

ebezzam committed:
Update reproducer, and dtype nits.

ebezzam reviewed (approved):
Thanks @lashahub and @Sreyan88 for the great work, and @eustlb and @ArthurZucker for the feedback 🤗 

Merging!

ebezzam committed:
Merge branch 'main' into audioflamingo3
